matrix_sdk/src/client.rs

```diff
@@ -679,8 +714,205 @@ impl Client {
+    /// [`get_sso_login_url`] and [`login_with_token`] directly.
+    ///
+    /// This should only be used for the first login, note that if the device id
+    /// isn't provided a new device will be created.
+    ///
+    /// If this isn't the first login, a device id should be provided to restore
+    /// the correct stores.
+    ///
+    /// The [`restore_login`] method should be used to restore a
+    /// logged in client after the first login.
+    ///
+    /// # Arguments
+    ///
+    /// * `server_url` - The local URL the server is going to try to bind to, e.g.
+    ///     `http://127.0.0.1:3030`.
+    ///
+    /// * `device_id` - A unique id that will be associated with this session. If
+    ///     not given the homeserver will create one. Can be an existing device_id
+    ///     from a previous login call. Note that this should be provided only
+    ///     if the client also holds the encryption keys for this device.
+    ///
+    /// * `initial_device_display_name` - A public display name that will be
+    ///     associated with the device_id. Only necessary the first time you
+    ///     login with this device_id. It can be changed later.
+    ///
+    /// # Example
+    /// ```no_run
+    /// # use std::convert::TryFrom;
+    /// # use matrix_sdk::Client;
+    /// # use matrix_sdk::identifiers::DeviceId;
+    /// # use matrix_sdk_common::assign;
+    /// # use futures::executor::block_on;
+    /// # use url::Url;
+    /// # let homeserver = Url::parse("https://example.com").unwrap();
+    /// # block_on(async {
+    /// let client = Client::new(homeserver).unwrap();
+    ///
+    /// let response = client
+    ///     .login_with_sso(None, Some("My app"))
+    ///     .unwrap();
+    ///
+    /// println!("Logged in as {}, got device_id {} and access_token {}",
+    ///          response.user_id, response.device_id, response.access_token);
+    /// # })
+    /// ```
+    ///
+    /// [`get_sso_login_url`]: #method.get_sso_login_url
+    /// [`login_with_token`]: #method.login_with_token
+    /// [`restore_login`]: #method.restore_login
+    #[cfg(feature = "sso_login")]
+    #[cfg_attr(feature = "docs", doc(cfg(sso_login)))]
+    pub async fn login_with_sso(
+        &self,
+        server_url: &str,
+        device_id: Option<&str>,
+        initial_device_display_name: Option<&str>,
+    ) -> Result<login::Response> {
+        info!("Logging in to {}", self.homeserver);
+        let url = Url::parse(server_url).unwrap();
+
+        let server = match tiny_http::Server::http((url.host_str().unwrap(), url.port().unwrap())) {
+            Ok(s) => s,
+            Err(err) => {
+                return Err(IoError::new(
+                    IoErrorKind::Other,
+                    format!("Could not spawn server: {}", err),
+                )
+                .into())
+            }
+        };
+        let sso_url = self.get_sso_login_url(server_url)?;
+
+        let exit_status = open::that(sso_url.to_string())?;
```

Comment:
I wonder if we should instead of automatically opening the URL here pass in a callback to the `login_with_sso()` method which gets called with the URL as an argument (I think you suggested that as well somewhere).

This would of course move the responsibility of opening the browser to the caller but it would also make stuff more flexible and remove our dependency on the `open` crate.

Why we would want more flexibility here? Well Weechat for example only will want to print out the SSO URL so people can open the browser on their local machine if they are using Weechat over SSH/Tmux.

I imagine some clients might also want to use this SSO login method if they have their own in-client web view and open the URL in that web view.

This would also make it easier to produce tests for this.

What do you think?

Reply:
I believe you are right. I wanted to make the callback optional and use `open` by default but it's probably better to do it how you suggest and remove the dependency. For example in Fractal we have a GIO function that does this so we don't need the dependency either.

Reply:
Oh nice, if even Fractal doesn't want `open`, let's use a callback instead.